napari/layers/utils/color_manager.py

```diff
@@ -0,0 +1,521 @@
+            is colormap or cycle.
+        """
+        if self.mode == ColorMode.DIRECT:
+            self.colors = np.concatenate(
+                (self.colors, transform_color(colors))
+            )
+        else:
+            color_property_name = self.color_properties.name
+            current_value = self.color_properties.current_value
+            old_properties = self.color_properties.values
+            values_to_add = properties[color_property_name]
+            new_color_property_values = np.concatenate(
+                (old_properties, values_to_add),
+                axis=0,
+            )
+
+            self.color_properties = ColorProperties(
+                name=color_property_name,
+                values=new_color_property_values,
+                current_value=current_value,
+            )
+
+    def update_current_properties(
+        self, current_properties: Dict[str, np.ndarray]
+    ):
+        """This is updates the current_value of the color_properties when the
+        layer current_properties is updated.
+
+        This is a convenience method that is generally only called by the layer.
+
+        Parameters
+        ----------
+        current_properties : Dict[str, np.ndarray]
+            The new current property values
+        """
+        if self.color_properties is not None:
+            current_property_name = self.color_properties.name
+            current_property_values = self.color_properties.values
+            if current_property_name in current_properties:
+                new_current_value = np.squeeze(
+                    current_properties[current_property_name]
+                )
+                if new_current_value != self.color_properties.current_value:
+                    self.color_properties = ColorProperties(
+                        name=current_property_name,
+                        values=current_property_values,
+                        current_value=new_current_value,
+                    )
+
+    def update_current_color(
+        self, current_color: np.ndarray, update_indices: list = []
+    ):
+        """Update the current color and update the colors if requested.
+
+        This is a convenience method and is generally called by the layer.
+
+        Parameters
+        ----------
+        current_color : np.ndarray
+            The new current color value.
+        update_indices : list
+            The indices of the color elements to update.
+            If the list has length 0, no colors are updated.
+            If the ColorManager is not in DIRECT mode, updating the values
+            will change the mode to DIRECT.
+        """
+        self.current_color = transform_color(current_color)[0]
+        if len(update_indices) > 0:
+            self.mode = ColorMode.DIRECT
+            cur_colors = self.colors.copy()
+            cur_colors[update_indices] = self.current_color
+            self.colors = cur_colors
+
+    @classmethod
+    def from_layer_kwargs(
```

Comment:
maybe make this private, as after more refactoring it might not be needed?

Reply:
Done!